lib/uxtx.js

```diff
@@ -0,0 +1,307 @@
+        break;
+      case this.TYPES.unknown:
+      default:
+        // something went wrong...
+        coins = [];
+    }
+
+    const value = coins.reduce((a, c) => a + c.value, 0);
+    const amount = new Amount(value, 'sat');
+
+    return amount.to(unit);
+  }
+
+  getJSON() {
+    assert(this._json);
+    return this._json;
+  }
+
+  // not required, no validation
+  getWallet() {
+    return this._wallet;
+  }
+  // not required, no validation
+  getAccount() {
+    return this._account;
+  }
+
+  getUXType() {
+    if (this._UXType)
+      return this._UXType;
+
+    let UXType;
+    if (this.isDeposit())
+      UXType = this.TYPES.deposit;
+
+    else if (this.isWithdraw())
+      UXType = this.TYPES.withdraw;
+
+    else if (this.isCoinbase())
+      UXType = this.TYPES.coinbase;
+
+    else
+      UXType = this.TYPES.unknown;
+
+    this._UXType = UXType;
+    return UXType;
+
+  }
+
+  // TODO: remove COPY suffix if in copy object
+  // TODO: parse accounts
+  /*
+  accounts = knownOutputs
+    .reduce((a, o) => ([...a, { name: o.name, account: o.account } ]), []);
+  */
+
+  getCounterparty() {
+    let counterparty;
+    let recipients;
+
+    const copy = this.getCopy();
+    const json = this.getJSON();
+
+    switch(this.getUXType()) {
+      case this.TYPES.deposit: // receive transaction
+        // parse counterparty
+        console.log('deposit')
+        if (json.inputs.length > 1)
+          counterparty = copy.MULTIPLE_ADDRESS_COPY;
+        else if (json.inputs.length === 1)
+          counterparty = json.inputs[0].address;
+        else
+          counterparty = copy.UNNOWN_ADDRESS_COPY;
+
+        // parse recipients
+        // TODO: catch edge cases here around more complex txns
+        recipients = json.inputs.map(i => i.address)
+
+        break;
+      case this.TYPES.withdraw: // send transaction
+        // parse recipients
+        recipients = json.outputs
+          .filter(o => {
+            // only will have path property
+            // if controlled by wallet
+            if (o.path)
+              return !o.path.change;
+              // handle sending to controlled wallet
+
+            return true; // all remaining outputs included
+          })
+          .map(o => o.address);
+        // NOTE: if need more information about
+        // coin, remove map above
+
+        // parse counterpary based on number of recipients
+        if (recipients.length === 1)
+          counterparty = recipients[0];
+        else if (recipients.length > 1)
+          counterparty = copy.MULTIPLE_ADDRESS_COPY;
+        else
+          counterparty = copy.UNKNOWN_ADDRESS_COPY;
+
+        break;
+      case this.TYPES.coinbase:
+        // assume one input and one output
+        counterparty = json.outputs[0].address;
+        recipients = [json.outputs[0].address];
+
+        break;
+      case this.TYPES.unknown:
+      default:
+        counterparty = copy.UNKNOWN_ADDRESS_COPY;
+        recipients = [];
+    }
+
```

Comment:
Maybe do a check to not overwrite the cache if it already exists?
All of the data is parsed here to calculate `recipients`, so assign it but
need to be careful about bugs here if there are multiple methods that calculate `this._recipients`...

Maybe should only have one method that calculates recipients and do extra work to maintain simplicity
until performance becomes a problem